File: Python/Combinatorics/test_combinatorics.py
"""
test_combinatorics.py
Author: Андрій Будильников

Test file for combinatorics module
"""

from combinatorics import factorial, permutation, combination, catalan_number

def test_factorial():
    """test factorial function"""
    assert factorial(0) == 1
    assert factorial(1) == 1
    assert factorial(5) == 120
    assert factorial(10) == 3628800
    print("factorial tests passed")

def test_permutation():
    """test permutation function"""
    assert permutation(5, 0) == 1
    assert permutation(5, 1) == 5
    assert permutation(5, 3) == 60
    assert permutation(10, 2) == 90
    print("permutation tests passed")

def test_combination():
    """test combination function"""
    assert combination(5, 0) == 1
    assert combination(5, 1) == 5
    assert combination(5, 3) == 10
    assert combination(10, 2) == 45
    print("combination tests passed")

def test_catalan_number():
    """test catalan number function"""
    # first few catalan numbers: 1, 1, 2, 5, 14, 42, 132, 429, 1430, 4862
    assert catalan_number(0) == 1
    assert catalan_number(1) == 1
    assert catalan_number(2) == 2
    assert catalan_number(3) == 5
    assert catalan_number(4) == 14
    assert catalan_number(5) == 42
    print("catalan number tests passed")

if __name__ == "__main__":
    test_factorial()
    test_permutation()
    test_combination()
    test_catalan_number()
    print("all tests passed! 🎉")

File: C++/Combinatorics/test_combinatorics.cpp
/*
 * test_combinatorics.cpp
 * Author: Андрій Будильников
 * 
 * Test program for combinatorics functions
 */

#include <iostream>
#include "Combinatorics.h"

int main() {
    std::cout << "Testing combinatorics functions...\n\n";
    
    // test factorial
    std::cout << "Factorial(0) = " << Combinatorics::factorial(0) << std::endl;
    std::cout << "Factorial(1) = " << Combinatorics::factorial(1) << std::endl;
    std::cout << "Factorial(5) = " << Combinatorics::factorial(5) << std::endl;
    std::cout << "Factorial(10) = " << Combinatorics::factorial(10) << std::endl;
    
    std::cout << std::endl;
    
    // test permutation
    std::cout << "Permutation(5, 0) = " << Combinatorics::permutation(5, 0) << std::endl;
    std::cout << "Permutation(5, 1) = " << Combinatorics::permutation(5, 1) << std::endl;
    std::cout << "Permutation(5, 3) = " << Combinatorics::permutation(5, 3) << std::endl;
    std::cout << "Permutation(10, 2) = " << Combinatorics::permutation(10, 2) << std::endl;
    
    std::cout << std::endl;
    
    // test combination
    std::cout << "Combination(5, 0) = " << Combinatorics::combination(5, 0) << std::endl;
    std::cout << "Combination(5, 1) = " << Combinatorics::combination(5, 1) << std::endl;
    std::cout << "Combination(5, 3) = " << Combinatorics::combination(5, 3) << std::endl;
    std::cout << "Combination(10, 2) = " << Combinatorics::combination(10, 2) << std::endl;
    
    std::cout << std::endl;
    
    // test catalan number
    std::cout << "CatalanNumber(0) = " << Combinatorics::catalanNumber(0) << std::endl;
    std::cout << "CatalanNumber(1) = " << Combinatorics::catalanNumber(1) << std::endl;
    std::cout << "CatalanNumber(2) = " << Combinatorics::catalanNumber(2) << std::endl;
    std::cout << "CatalanNumber(3) = " << Combinatorics::catalanNumber(3) << std::endl;
    std::cout << "CatalanNumber(4) = " << Combinatorics::catalanNumber(4) << std::endl;
    std::cout << "CatalanNumber(5) = " << Combinatorics::catalanNumber(5) << std::endl;
    
    std::cout << "\nAll tests completed! 🎉" << std::endl;
    
    return 0;
}